src/v/pandaproxy/schema_registry/json.cc

```diff
@@ -1159,79 +1346,119 @@ bool is_object_dependencies_superset(
+                  std::move(path_dep),
+                  json_incompatibility_type::dependency_schema_added);
+                return;
+            }
 
-              auto const& n = n_it->value;
+            // schemas: o and n needs to be compatible
+            res.merge(is_superset(ctx, o, n, std::move(path_dep)));
+        } else if (o.IsArray()) {
+            if (n_it == newer_p.MemberEnd()) {
+                res.emplace<json_incompatibility>(
+                  std::move(path_dep),
+                  json_incompatibility_type::dependency_array_added);
+                return;
+            }
 
-              if (!n.IsArray()) {
-                  return false;
-              }
-              // string array: n needs to be a a superset of o
-              // TODO: n^2 search
-              bool n_superset_of_o = std::ranges::all_of(
-                o.GetArray(), [n_array = n.GetArray()](json::Value const& p) {
-                    return std::ranges::find(n_array, p) != n_array.End();
-                });
-              if (!n_superset_of_o) {
-                  return false;
-              }
-              return true;
-          } else {
-              throw as_exception(invalid_schema(fmt::format(
-                "dependencies can only be an array or an object for valid "
-                "schemas but it was: {}",
-                pj{o})));
-          }
-      });
+            auto const& n = n_it->value;
+
+            if (!n.IsArray()) {
+                res.emplace<json_incompatibility>(
+                  std::move(path_dep),
+                  json_incompatibility_type::dependency_array_added);
+                return;
+            }
+            // string array: n needs to be a a superset of o
+            // TODO: n^2 search
+            bool n_superset_of_o = std::ranges::all_of(
+              o.GetArray(), [n_array = n.GetArray()](json::Value const& p) {
+                  return std::ranges::find(n_array, p) != n_array.End();
+              });
+            if (!n_superset_of_o) {
+                bool o_superset_of_n = std::ranges::all_of(
+                  n.GetArray(), [o_array = o.GetArray()](json::Value const& p) {
+                      return std::ranges::find(o_array, p) != o_array.End();
+                  });
+                if (o_superset_of_n) {
+                    res.emplace<json_incompatibility>(
+                      std::move(path_dep),
+                      json_incompatibility_type::dependency_array_extended);
+                } else {
+                    res.emplace<json_incompatibility>(
+                      std::move(path_dep),
+                      json_incompatibility_type::dependency_array_changed);
+                }
+            }
+            return;
+        } else {
+            throw as_exception(invalid_schema(fmt::format(
+              "dependencies can only be an array or an object for valid "
+              "schemas but it was: {}",
+              pj{o})));
+        }
```

Comment:
nit: json schema validation should prevent reaching this block, but at the same time it's not harmful to leave it here.